pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1704,3 +1706,68 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta PredicateMetadata, no
 	klog.V(5).Infof("All PVCs found matches for pod %v/%v, node %q", pod.Namespace, pod.Name, node.Name)
 	return true, nil, nil
 }
+
+// EvenPodsSpreadPredicate checks if a pod can be scheduled on a node which satisfies
+// its topologySpreadConstraints.
+func EvenPodsSpreadPredicate(pod *v1.Pod, meta PredicateMetadata, nodeInfo *schedulernodeinfo.NodeInfo) (bool, []PredicateFailureReason, error) {
+	node := nodeInfo.Node()
+	if node == nil {
+		return false, nil, fmt.Errorf("node not found")
+	}
+
+	constraints := getHardTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return true, nil, nil
+	}
+
+	var topologyPairsPodSpreadMap *topologyPairsPodSpreadMap
+	if predicateMeta, ok := meta.(*predicateMetadata); ok {
+		topologyPairsPodSpreadMap = predicateMeta.topologyPairsPodSpreadMap
+	} else { // We don't have precomputed metadata. We have to follow a slow path to check spread constraints.
+		// TODO(Huang-Wei): get it implemented
```

Comment:
Is there actually a legitimate case where metadata is not pre-computed and we still would like to proceed and compute it here? If yes, can you please document that case here. 

I hope that we can always consider this an error, and not invoke the metadata generation logic here.

Reply:
This always puzzled me as well. I once asked @bsalamat on this, the answer i got is that ClusterAutoscaler seems to use a non-default algorithmProvider, so they are the only consumer of the else block.

To be honest, I haven't checked the exact running path of CusterAutoscaler, and also have no knowledge on their background and usecase.

Reply:
can you please create an issue to track this? we also need an issue to track optimizing the implementation by using counters instead of pod lists in the maps.

Reply:
Let me put them as TODO items in the umbrella issue for now, and will create Issues/PRs resolving them after the code gets merged.